packages/app/src/server/routes/api/cortex/cortexRuntimeConfig.ts

```diff
@@ -0,0 +1,235 @@
+    await validateAndExtractRuntimeConfig(req.body);
+  } catch (err) {
+    // Will structure the error for the client in a consistent struct and will log a consistent error log structure
+    // with the message from failed validation
+    return next(new GeneralServerError(400, `bad request: ${err.message}`));
+  }
+
+  if (KUBECONFIG === undefined) {
+    // This will be caught and logged with a stack trace and handled correctly by our error middleware,
+    // issueing a 500 back to the user with the error message below in the resonse body
+    throw new Error("internal error: kubeconfig not set");
+  }
+
+  // Assume that `req.body` is the new config map's payload content.
+  const newcm = genCortexRuntimeConfigCM(KUBECONFIG, req.body);
+  try {
+    await createOrUpdateConfigMapWithRetry(newcm, true);
+  } catch (err) {
+    // Expected error for invalid documents: response status
+    // 422 Unprocessable Entity
+    if (err.response !== undefined) {
+      // Reflecting the HTTP response returned by the k8s API. Example:
+      // 2021-05-19 18:33:24
+      // 2021-05-19T16:33:24.678Z warning: e.response: {
+      // 2021-05-19 18:33:24
+      //   "statusCode": 422,
+      // 2021-05-19 18:33:24
+      //   "body": {
+      // 2021-05-19 18:33:24
+      //     "kind": "Status",
+      // 2021-05-19 18:33:24
+      //     "apiVersion": "v1",
+      // 2021-05-19 18:33:24
+      //     "metadata": {},
+      // 2021-05-19 18:33:24
+      //     "status": "Failure",
+      const kr = err.response;
+      // forward status code and log the message with the error logging middleware
+      return next(
+        new GeneralServerError(
+          kr.statusCode,
+          `Error during cortex runtime config map update. ` +
+            `Kubernetes API returned status code ${kr.statusCode}. ` +
+            `Message: ${kr.body.message}`
+        )
+      );
+    }
+
+    // When no HTTP response was received, i.e. upon transient errors.
+    log.warning("error during config map update: %s", err); // log with stack trace
+    return next(
+      new GeneralServerError(
+        500,
+        `error during config map update: ${err.message}`
+      )
+    );
+  }
+
+  res.status(202).send("accepted: change is expected to take effect soon");
+  return;
+}
+
+/**
+ * Reads and returns the Cortex runtime config as specified in the ConfigMap.
+ * This is the source of truth for runtime config. Cortex provides the "observed"
+ * runtime config through it's API, but the value of that can lag because
+ * Cortex will poll for changes in this file periodically to refresh
+ * the "observed" runtime config.
+ */
+export async function readCortexRuntimeConfigHandler(
+  req: Request,
+  res: Response,
+  next: NextFunction
+) {
+  if (KUBECONFIG === undefined) {
+    // This will be logged with a stack trace and handled correctly by our error middleware,
+    // issueing a 500 back to the user with the error message below in the resonse body
```

Comment:
Thanks for that comment -- sure that "kubeconfig not set" does end up being written to the HTTP client? In the outest and most general error handler in a web framework one usually does not want to leak error detail to the client, but only a generic error message like "internal server error"; also for security reasons.

Reply:
Yes, this will be written to the HTTP client! In theory, we should never hit this issue because we load the KUBECONFIG at process startup, and we exit(1) if it fails, so the pod would crash loop.

I wasn't concerned with this particular error leaking to the client, quite the opposite - as a System Admin, I'd like to see the error quickly, especially when all this code is open and visible anyway. What would the security concern be in this case?